targets/simple_switch_grpc/tests/test_grpc_counter.cpp

```diff
@@ -0,0 +1,142 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc++/grpc++.h>
+
+#include <p4/bm/dataplane_interface.grpc.pb.h>
+#include <p4/p4runtime.grpc.pb.h>
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <string>
+
+#include "base_test.h"
+
+namespace sswitch_grpc {
+
+namespace testing {
+
+namespace {
+
+constexpr char counter_json[] = TESTDATADIR "/counter.json";
+constexpr char counter_proto[] = TESTDATADIR "/counter.proto.txt";
+
+class SimpleSwitchGrpcTest_Counter : public SimpleSwitchGrpcBaseTest {
+ protected:
+  SimpleSwitchGrpcTest_Counter()
+      : SimpleSwitchGrpcBaseTest(counter_proto),
+        dataplane_channel(grpc::CreateChannel(
+            dp_grpc_server_addr, grpc::InsecureChannelCredentials())),
+        dataplane_stub(p4::bm::DataplaneInterface::NewStub(
+            dataplane_channel)) { }
+
+  void SetUp() override {
+    SimpleSwitchGrpcBaseTest::SetUp();
+    update_json(counter_json);
+  }
+
+  std::shared_ptr<grpc::Channel> dataplane_channel{nullptr};
+  std::unique_ptr<p4::bm::DataplaneInterface::Stub> dataplane_stub{nullptr};
+};
+
+TEST_F(SimpleSwitchGrpcTest_Counter, CounterHit) {
+  auto t_id = get_table_id(p4info, "t_redirect");
+  auto mf_id = get_mf_id(p4info, "t_redirect", "sm.packet_length");
+  auto a_id = get_action_id(p4info, "port_redirect");
+  auto dc_id = get_direct_counter_id(p4info, "cntr");
+
+  p4::Entity entity;
+  auto table_entry = entity.mutable_table_entry();
+  table_entry->set_table_id(t_id);
+  table_entry->set_priority(100);
+  auto match = table_entry->add_match();
+  match->set_field_id(mf_id);
+  auto exact = match->mutable_exact();
+  auto table_action = table_entry->mutable_action();
+  auto action = table_action->mutable_action();
+  action->set_action_id(a_id);
+
+  auto write_one_entry = [&, this](const std::string &key_string) {
+    p4::WriteRequest write_request;
+    write_request.set_device_id(device_id);
+    auto update = write_request.add_updates();
+    update->set_type(p4::Update_Type_INSERT);
+    exact->set_value(key_string);
+    *(update->mutable_entity()) = entity;
+
+    p4::WriteResponse write_response;
+    ClientContext context;
+    auto status = Write(&context, write_request, &write_response);
+    EXPECT_TRUE(status.ok());
+  };
+  write_one_entry(std::string("\x00\x00\x00\x0a", 4));
+  write_one_entry(std::string("\x00\x00\x00\x05", 4));
+
+  auto send_one_packet = [&, this](const std::string &payload) {
```

Comment:
same here, is there a reason for using automatic variable capture?

Reply:
Fixed